### PR TITLE
Code of "GCNet: Non-local Networks Meet Squeeze-Excitation Networks and Beyond"

### PR DESCRIPTION
**Pull request by xvjiarui**
Add gcb and configs

### CODE REVIEW

**Comment by xvjiarui** [comment]
It seems like two of config files are too long (>79). Any idea?

**Comment by hellock** [comment]
> It seems like two of config files are too long (>79). Any idea?

You can just add `# noqa: E501` at the end of the line to ignore the linting error.

File: mmdet/ops/gcb/context_block.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    else:
+        constant_init(m, val=0)
+
+
+class ContextBlock(nn.Module):
+
+    def __init__(self, inplanes, planes, pool, fusions):
+        super(ContextBlock, self).__init__()
+        assert pool in ['avg', 'att']
+        assert all([f in ['channel_add', 'channel_mul'] for f in fusions])
```
Comment (hellock) [inline mmdet/ops/gcb/context_block.py]: May `assert isinstance(fusions, (list, tuple))` first.

File: mmdet/ops/gcb/context_block.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        if self.channel_mul_conv is not None:
+            # [N, C, 1, 1]
+            channel_mul_term = torch.sigmoid(self.channel_mul_conv(context))
+            out = x * channel_mul_term
+        else:
+            out = x
+        if self.channel_add_conv is not None:
+            # [N, C, 1, 1]
+            channel_add_term = self.channel_add_conv(context)
+            out = out + channel_add_term
```
Comment (hellock) [inline mmdet/ops/gcb/context_block.py]: May be more clear:
```python
out = x
if self.channel_mul_conv is not None:
    channel_mul_term = self.channel_mul_conv(context).sigmoid()
    out *= channel_mul_term
if self.channel_add_conv is not None:
    channel_add_term = self.channel_add_conv(context)
    out += channel_add_term
```

File: mmdet/ops/gcb/context_block.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    def __init__(self, inplanes, planes, pool, fusions):
+        super(ContextBlock, self).__init__()
+        assert pool in ['avg', 'att']
+        assert all([f in ['channel_add', 'channel_mul'] for f in fusions])
+        assert len(fusions) > 0, 'at least one fusion should be used'
+        self.inplanes = inplanes
+        self.planes = planes
+        self.pool = pool
+        self.fusions = fusions
+        if 'att' in pool:
```
Comment (hellock) [inline mmdet/ops/gcb/context_block.py]: You mean `pool == 'att'`?

File: mmdet/ops/gcb/context_block.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+def last_zero_init(m):
+    if isinstance(m, nn.Sequential):
+        constant_init(m[-1], val=0)
+    else:
+        constant_init(m, val=0)
+
+
+class ContextBlock(nn.Module):
+
+    def __init__(self, inplanes, planes, pool, fusions):
```
Comment (hellock) [inline mmdet/ops/gcb/context_block.py]: May be better to add a suffix like `method` or `type`?
`pooling_type/pooling_method`
`fusion_types/fusion_methods`

File: mmdet/ops/gcb/context_block.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            # [N, 1, C, H * W]
+            input_x = input_x.unsqueeze(1)
+            # [N, 1, H, W]
+            context_mask = self.conv_mask(x)
+            # [N, 1, H * W]
+            context_mask = context_mask.view(batch, 1, height * width)
+            # [N, 1, H * W]
+            context_mask = self.softmax(context_mask)
+            # [N, 1, H * W, 1]
+            context_mask = context_mask.unsqueeze(3)
```
Comment (hellock) [inline mmdet/ops/gcb/context_block.py]: May be more safe by `context_mask.unsqueeze(-1)`

File: mmdet/ops/gcb/context_block.py (hunk 2 above)
Comment (xvjiarui) [reply mmdet/ops/gcb/context_block.py]: In-place operator is not applicable here. I used `out = out * channel_mul_term` and `out = out + channel_add_term` instead.

**Comment by xvjiarui** [comment]
Rename method -> type.
Config variables are updated to adapt master.

**Pull request merged**